test/test_bounds.cpp

```diff
@@ -0,0 +1,130 @@
+//------------------------------------------------------------------------------
+// © 2021-2024. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#include <cstdio>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-eos/base/fast-math/logs.hpp>
+#include <singularity-eos/base/table_bounds.hpp>
+
+#ifndef CATCH_CONFIG_FAST_COMPILE
+#define CATCH_CONFIG_FAST_COMPILE
+#include <catch2/catch_test_macros.hpp>
+#endif
+
+constexpr int NGRIDS = 3;
+using Bounds = singularity::Bounds<NGRIDS>;
+using RegularGrid1D = Spiner::RegularGrid1D<Real>;
+using Grid_t = Bounds::Grid_t;
+
+constexpr Real rho_normal = 3; // g/cm^3
+constexpr Real T_normal = 300; // Kelvin
+constexpr Real rho_min = 1e-6;
+constexpr Real rho_max = 1e3;
+constexpr Real T_min = 1;
+constexpr Real T_max = 1e9;
+constexpr Real T_split = 1e4;
+constexpr int N_per_decade_fine = 200;
+constexpr Real N_factor = 5;
+
+SCENARIO("Bounds can compute number of points from points per decade", "[Bounds]") {
+  WHEN("We compute the number of points from points per decade") {
+    int np = Bounds::getNumPointsFromPPD(T_min, T_max, N_per_decade_fine);
+    THEN("We get the right number") { REQUIRE(np == 1800); }
+  }
+}
+
+SCENARIO("Linear bounds in the bounds object", "[Bounds]") {
+  WHEN("We compute linear bounds ") {
+    constexpr Real min = -2;
+    constexpr Real max = 5;
+    constexpr int N = 21;
+    Bounds lin_bounds(min, max, N);
+    THEN("The min and max and point number correct") {
+      REQUIRE(lin_bounds.grid.min() == min);
+      REQUIRE(lin_bounds.grid.max() == max);
+      REQUIRE(lin_bounds.grid.nPoints() == N);
+    }
+  }
+}
+
+SCENARIO("Logarithmic, single-grid bounds in the bounds object", "[Bounds]") {
+  WHEN("We compute logarithmic single-grid bounds") {
+    int np = Bounds::getNumPointsFromPPD(rho_min, rho_max, N_per_decade_fine);
+    Bounds lRhoBounds(rho_min, rho_max, np, true, 0.0, rho_normal);
+    THEN("The lower and upper bounds are right") {
+      REQUIRE(std::abs(lRhoBounds.grid.min() - singularity::FastMath::log10(rho_min)) <=
+              1e-12);
```

Comment:
You use `1e-12` in a lot of places for a tolerance. Two comments:
1. I would at least like this to be a `constexpr` defined parameter since it's repeated so much
2. It seems like the value might be predicated on the type of `Real`. With the idea of making things flexible, would it make sense to define
    ```c++
    constexpr REAL_TOL = std::numeric_limits<Real>::epsilon * 1e3
    ```
    instead?

I'll let this be a blanket comment for the entire file rather than pointing out every instance.

Reply:
Done.